File: schemas/src/synesis_schemas/main_server/runs.py
import uuid
from typing import Literal, List, Optional, Union
from datetime import datetime, timezone
from pydantic import BaseModel


# DB Models


RUN_TYPE_LITERAL = Literal["swe", "analysis"]


class RunInDB(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    type: RUN_TYPE_LITERAL
    status: Literal["pending", "running", "completed", "failed", "rejected"]
    run_name: str
    project_id: Optional[uuid.UUID] = None
    conversation_id: Optional[uuid.UUID] = None
    plan_and_deliverable_description_for_user: str
    plan_and_deliverable_description_for_agent: str
    questions_for_user: Optional[str] = None
    configuration_defaults_description: Optional[str] = None
    started_at: datetime
    completed_at: Optional[datetime] = None


class RunMessageInDB(BaseModel):
    id: uuid.UUID
    content: str
    run_id: uuid.UUID
    type:  Literal["tool_call", "result", "error"]
    created_at: datetime = datetime.now(timezone.utc)


class RunPydanticMessageInDB(BaseModel):
    id: uuid.UUID
    run_id: uuid.UUID
    message_list: bytes
    created_at: datetime = datetime.now(timezone.utc)


class DataSourceInRunInDB(BaseModel):
    run_id: uuid.UUID
    data_source_id: uuid.UUID
    created_at: datetime


class DatasetInRunInDB(BaseModel):
    run_id: uuid.UUID
    dataset_id: uuid.UUID
    created_at: datetime


class ModelEntityInRunInDB(BaseModel):
    run_id: uuid.UUID
    model_entity_id: uuid.UUID
    created_at: datetime


class PipelineInRunInDB(BaseModel):
    run_id: uuid.UUID
    pipeline_id: uuid.UUID
    created_at: datetime


class AnalysisInRunInDB(BaseModel):
    run_id: uuid.UUID
    analysis_id: uuid.UUID
    created_at: datetime


class AnalysisFromRunInDB(BaseModel):
    run_id: uuid.UUID
    analysis_id: uuid.UUID
    created_at: datetime


class PipelineFromRunInDB(BaseModel):
    run_id: uuid.UUID
    pipeline_id: uuid.UUID
    created_at: datetime


# API Models


class RunEntityIds(BaseModel):
    data_source_ids: List[uuid.UUID] = []
    dataset_ids: List[uuid.UUID] = []
    model_entity_ids: List[uuid.UUID] = []
    pipeline_ids: List[uuid.UUID] = []
    analysis_ids: List[uuid.UUID] = []


class Run(RunInDB):
    inputs: Optional[RunEntityIds] = None
    outputs: Optional[RunEntityIds] = None


class MessageForLog(BaseModel):
    content: str
    type: Literal["tool_call", "result", "error"]
    write_to_db: int = 1
    target: Literal["redis", "taskiq", "both"] = "both"
    created_at: datetime = datetime.now(timezone.utc)


class CodeForLog(BaseModel):
    code: str
    filename: str
    target: Literal["redis", "taskiq", "both"] = "both"
    # cant do None because redis
    output: str = ""
    error: str = ""
    created_at: datetime = datetime.now(timezone.utc)


# Create Models


class RunCreate(BaseModel):
    type: RUN_TYPE_LITERAL
    initial_status: Literal["pending", "running",
                            "completed", "failed"] = "pending"
    run_name: str
    plan_and_deliverable_description_for_user: str
    plan_and_deliverable_description_for_agent: str
    questions_for_user: Optional[str] = None
    configuration_defaults_description: Optional[str] = None
    target_entity_id: Optional[uuid.UUID] = None
    project_id: Optional[uuid.UUID] = None
    conversation_id: Optional[uuid.UUID] = None
    data_sources_in_run: List[uuid.UUID] = []
    datasets_in_run: List[uuid.UUID] = []
    model_entities_in_run: List[uuid.UUID] = []
    pipelines_in_run: List[uuid.UUID] = []
    analyses_in_run: List[uuid.UUID] = []


class RunMessageCreate(BaseModel):
    type: Literal["tool_call", "result", "error"]
    content: str
    run_id: uuid.UUID


class RunMessageCreatePydantic(BaseModel):
    content: bytes
    run_id: uuid.UUID


# Update Models

class RunStatusUpdate(BaseModel):
    run_id: uuid.UUID
    status: Literal["running", "completed", "failed", "rejected"]
    summary: Optional[str] = None

File: main_server/client/src/types/runs.ts
import { UUID } from "crypto";

export type RunType = "swe" | "analysis" | "extraction";

// DB Models

export interface RunInDB {
  id: UUID;
  userId: UUID;
  type: RunType;
  status: "pending" | "running" | "completed" | "failed" | "rejected";
  runName: string;
  projectId?: UUID | null;
  conversationId?: UUID | null;
  planAndDeliverableDescriptionForUser: string;
  planAndDeliverableDescriptionForAgent: string;
  questionsForUser?: string | null;
  configurationDefaultsDescription?: string | null;
  startedAt: string;
  completedAt?: string | null;
}

export interface RunMessageInDB {
  id: UUID;
  content: string;
  runId: UUID;
  type: "tool_call" | "result" | "error";
  createdAt: string;
}

export interface RunPydanticMessageInDB {
  id: UUID;
  runId: UUID;
  messageList: Uint8Array; // bytes in Python
  createdAt: string;
}

export interface DataSourceInRunInDB {
  runId: UUID;
  dataSourceId: UUID;
  createdAt: string;
}

export interface DatasetInRunInDB {
  runId: UUID;
  datasetId: UUID;
  createdAt: string;
}

export interface ModelEntityInRunInDB {
  runId: UUID;
  modelEntityId: UUID;
  createdAt: string;
}

export interface PipelineInRunInDB {
  runId: UUID;
  pipelineId: UUID;
  createdAt: string;
}

export interface AnalysisInRunInDB {
  runId: UUID;
  analysisId: UUID;
  createdAt: string;
}

export interface AnalysisFromRunInDB {
  runId: UUID;
  analysisId: UUID;
  createdAt: string;
}

export interface PipelineFromRunInDB {
  runId: UUID;
  pipelineId: UUID;
  createdAt: string;
}

// API Models

export interface RunEntityIds {
  dataSourceIds: UUID[];
  datasetIds: UUID[];
  modelEntityIds: UUID[];
  pipelineIds: UUID[];
  analysisIds: UUID[];
}

export interface Run extends RunInDB {
  inputs?: RunEntityIds | null;
  outputs?: RunEntityIds | null;
}

export interface StreamedCode {
  code: string;
  filename: string;
  target: "redis" | "taskiq" | "both";
  output: string;
  error: string;
  createdAt: string;
}

// Create Models

export interface RunCreate {
  type: RunType;
  initialStatus: "pending" | "running" | "completed" | "failed";
  runName: string;
  planAndDeliverableDescriptionForUser: string;
  planAndDeliverableDescriptionForAgent: string;
  questionsForUser?: string | null;
  configurationDefaultsDescription?: string | null;
  targetEntityId?: UUID | null;
  projectId?: UUID | null;
  conversationId?: UUID | null;
  dataSourcesInRun: UUID[];
  datasetsInRun: UUID[];
  modelEntitiesInRun: UUID[];
  pipelinesInRun: UUID[];
  analysesInRun: UUID[];
}

export interface RunMessageCreate {
  type: "tool_call" | "result" | "error";
  content: string;
  runId: UUID;
}

export interface RunMessageCreatePydantic {
  content: Uint8Array; // bytes in Python
  runId: UUID;
}

// Update Models

export interface RunStatusUpdate {
  runId: UUID;
  status: "running" | "completed" | "failed" | "rejected";
  summary?: string | null;
}
